stablehlo/tests/ops_stablehlo.mlir

```diff
@@ -2250,55 +2250,15 @@ func.func @slice(%arg0: tensor<3x4xi32>) -> tensor<1x2xi32> {
 
 // -----
 
-// CHECK-LABEL: func @slice
-func.func @slice_dynamic_dim(%arg0: tensor<3x?xi32>) -> tensor<1x?xi32> {
-  %0 = "stablehlo.slice"(%arg0) {
-    start_indices = dense<[1, 1]> : tensor<2xi64>,
-    limit_indices = dense<[2, 2]> : tensor<2xi64>,
-    strides = dense<[1, 1]> : tensor<2xi64>
-  } : (tensor<3x?xi32>) -> tensor<1x?xi32>
-  func.return %0 : tensor<1x?xi32>
-}
-
-// -----
-
-func.func @slice_dynamic_dim_invalid_indices(%arg0: tensor<3x?xi32>) -> tensor<1x?xi32> {
-  // expected-error@+1 {{negative start index -1 in dimension 1}}
-  %0 = "stablehlo.slice"(%arg0) {
-    start_indices = dense<[1, -1]> : tensor<2xi64>,
-    limit_indices = dense<[2, 2]> : tensor<2xi64>,
-    strides = dense<[1, 1]> : tensor<2xi64>
-  } : (tensor<3x?xi32>) -> tensor<1x?xi32>
-  func.return %0 : tensor<1x?xi32>
-}
-
-// -----
-
-// CHECK-LABEL: func @slice_unranked
-func.func @slice_unranked(%arg0: tensor<*xi32>) -> tensor<*xi32> {
-  %0 = "stablehlo.slice"(%arg0) {start_indices = dense<[1, 0]> : tensor<2xi64>, limit_indices = dense<[2, 4]> : tensor<2xi64>, strides = dense<[1, 2]> : tensor<2xi64>} : (tensor<*xi32>) -> tensor<*xi32>
-  func.return %0 : tensor<*xi32>
-}
-
-// -----
-
-func.func @slice_indices_mismatch(%arg0: tensor<3x4xi32>) -> tensor<1x4xi32> {
-  // expected-error@+1 {{failed to verify that all of {start_indices, limit_indices, strides} have same type}}
-  %0 = "stablehlo.slice"(%arg0) {start_indices = dense<[1, 2]> : tensor<2xi64>, limit_indices = dense<[2, 4, 1]> : tensor<3xi64>, strides = dense<[1, 2]> : tensor<2xi64>} : (tensor<3x4xi32>) -> tensor<1x4xi32>
-  func.return %0 : tensor<1x4xi32>
-}
-
-// -----
-
-func.func @slice_operand_result_mismatch(%arg0: tensor<3x4xi32>) -> tensor<1x4xf32> {
+func.func @slice_C1(%arg0: tensor<3x4xi32>) -> tensor<1x4xf32> {
```

Comment:
Another policy question - are we targeting complete coverage of constraints in these tests? If yes, then we'll want to have more tests e.g. in C2, C3. If no, how do we track that work? cc @sdasgup3

Reply:
>> are we targeting complete coverage of constraints in these tests
Yes, that is right except for the checks enforced by the ODS using Type/Attr constraints and traits.
The idea was to test the code we introduced in the cpp functions and skip testing the mlir specific code implementing the ODS constraints.

On a side note: For this op, the constraint C2 is split between ODS and cpp files. We have a test,  labeled `slice_C2`, for only the part tested in cpp file.

Reply:
As @zhouxin913 pointed out to me that
https://github.com/openxla/stablehlo/blob/main/docs/type_inference.md#p4-establish-testing-guidelines  seems contradictory to what I am proposing here. I will get back to that asap.

Reply:
"are we targeting complete coverage of constraints in these tests? Yes, that is right except for the checks enforced by the ODS using Type/Attr constraints and traits". Okay, then I have the following feedback. (And I'll also include my thought process around this).

We have the following constraints in the spec:

```
// Ix constraints come from the Inputs table
(I1) operand tensor
(I2) start_indices 1-dimensional tensor constant of type si64
(I3) limit_indices 1-dimensional tensor constant of type si64
(I4) strides 1-dimensional tensor constant of type si64
(C1) operand and result have the same element type.
(C2) size(start_indices) = size(limit_indices) = size(strides) = rank(operand).
(C3) 0 <= start_indices[d] <= limit_indices[d] <= dim(operand, d) for all dimension d.
(C4) 0 < strides[d] for all dimension d.
(C5) dim(result, d) = ... for all dimension d in operand.
```

From what I can see, these constraints will be comprehensively covered by the following tests:

```
I1: a) operand is not a tensor. (Covered by ODS).
I2: a) start_indices is not a tensor. (Covered by ODS).
    b) rank(start_indices) != 1.
    c) element_type(start_indices) != si64. (Covered by ODS).
I3: a) limit_indices is not a tensor. (Covered by ODS).
    b) rank(limit_indices) != 1. (Covered by ODS).
    c) element_type(limit_indices) != si64. (Covered by ODS).
I4: a) strides is not a tensor. (Covered by ODS).
    b) rank(strides) != 1. (Covered by ODS).
    c) element_type(strides) != si64. (Covered by ODS).
C1: element_type(operand) != element_type(result). (Covered by ODS).
C2: a) size(start_indices) != rank(operand).
    b) size(limit_indices) != rank(operand). (Covered by ODS).
    c) size(strides) != rank(operand). (Covered by ODS).
C3: a) start_indices < 0.
    b) start_indices > limit_indices.
    c) limit_indices > shape(operand).
C4: a) strides <= 0.
C5: a) limit_indices[d]-start_indices[d] is a multiple of strides[d].
    b) is not a multiple of strides[d].
```

If we drop the "Covered by ODS" pieces, this will leave us with the following test cases:

```
I2b: rank(start_indices) != 1.
C2a: size(start_indices) != rank(operand).
C3a: start_indices < 0.
C3b: start_indices > limit_indices.
C3c: limit_indices > shape(operand).
C4a: strides <= 0.
C5a: limit_indices[d]-start_indices[d] is a multiple of strides[d].
C5b: is not a multiple of strides[d].
```

Comparing these test cases with the actual test cases results in the following feedback on my part:
  1) The test for @slice_c1 doesn't seem to be needed (covered by ODS)
  2) The second test for @slice_c3 ("negative start index -1 in dimension 1") is redundant (covered by the first test).
  3) The test for @slice_indices_mismatch doesn't seem to be needed (covered by ODS).
  4) The test for C5b is missing.
  5) Let's not worry about dynamic shapes for now, since they aren't part of the spec at the moment.

Reply:
"Let's not worry about dynamic shapes for now". What I mean by this is "let's keep whatever tests for dynamic shapes that we have but not worry about comprehensively covering dynamism support for specific ops".

Reply:
Thanks for the detailed feedback! The brevity of breaking down constraints definitely helps see which negative tests are needed. It seems we agree on leaving constraints covered by the ODS and instead have a comment next to it. Also it's nice to point out that we should also cover requirements listed in the Inputs section.

Reply:
"Also it's nice to point out that we should also cover requirements listed in the Inputs section". If we're in alignment on the proposed strategy, let's also update the checklist. Also cc @sdasgup3.

Reply:
Agreed about C5b. All other tests check out as well, except that the first slice_c2 test is actually a slice_i2 test (noted in a separate comment thread).